typing/typecore.ml

```diff
@@ -5199,8 +5531,19 @@ and type_cases
   in
   let val_cases, exn_cases =
     match category with
-      | Value -> (cases : value case list), []
-      | Computation -> split_cases env cases in
+    | Value ->
+        let val_cases =
+          List.map
+            (fun htc -> { htc.untyped_case with pattern = htc.typed_pat })
```

Comment:
What's going on here? Where does this come from in the old code?

Reply:
This code corresponds to the old implementation of splitting value/exception cases of pattern matches. This PR doesn't change anything about the semantics of this check, but it does change the code structure. That's because the result of this splitting is now used in two places:
  1. once in `map_half_typed_cases` regardless of whether the case is for the last parameter of a function. That's so we can do a few checks: whether the match is total, whether variables are unused.
  2. once in `type_cases` where we know that a case is for the last parameter of a function and so has a expression body. The expression body is needed to perform the "ambiguous binding" check corresponding to warning 57, which can fire in the presence of `when` guards.

The version of the code you left the comment on runs the case-splitting twice. I didn't like this very much, so I've pushed a new commit that runs the case-splitting only once (but that runs some additional checks that the caller specifies, so that the `type_cases` caller can check for warning 57).